### PR TITLE
IGNITE-13558-4

### PR DESCRIPTION
**Pull request by denis-chudov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 1]
```diff
@@ -5519,66 +5519,76 @@ private void restorePartitionStates(
...
 
-                            topPartRef.updateAndGet(top0 -> {
-                                if (top0 == null)
-                                    return top;
+                                Comparator<T3<Long, Long, GroupPartitionId>> cmp = processedPartitionComparator();
 
-                                for (T3<Long, Long, GroupPartitionId> t2 : top0) {
-                                    top.add(t2);
+                                threadCtx.topPartRef.updateAndGet(top0 -> {
+                                    if (threadCtx.topPartRef.get() == null ||
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:5556]: Use `prev == null`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 2]
```diff
@@ -5519,66 +5519,76 @@ private void restorePartitionStates(
...
+                                        cmp.compare(threadCtx.topPartRef.get().last(), curPart) < 0) {
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:5557]: `threadCtx.topPartRef.get().last()` -> `prev.last()`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 3]
```diff
@@ -5519,66 +5519,76 @@ private void restorePartitionStates(
...
+
+                                        trimToSize(top, topPartRefLimit);
+
+                                        return top;
+                                    }
+                                    else
+                                        return top0;
+                                });
+
+                                RestorePartitionStateThreadContext.PROCESSED_CNT_UPD.incrementAndGet(threadCtx);
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:5573]: Let's make it a `RestorePartitionStateThreadContext` method.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManager.java [hunk 4]
```diff
@@ -89,16 +88,28 @@
...
 
+    /**
+     * Pre-create single partition that resides in page memory or WAL and restores their state.
+     *
+     * @param p Partition id.
+     * @param recoveryState Partition recovery state.
+     * @return Processing time in millis.
+     * @throws IgniteCheckedException If failed.
+     */
+    long restoreStateOfPartition(int p, @Nullable Integer recoveryState) throws IgniteCheckedException;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManager.java:99]: too similar to `restorePartitionStates`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManager.java [hunk 5]
```diff
@@ -89,16 +88,28 @@
...
-    Map<Integer, Long> restorePartitionStates(
-        Map<GroupPartitionId, Integer> partRecoveryStates
-    ) throws IgniteCheckedException;
+    void restorePartitionStates() throws IgniteCheckedException;
+
+    /**
+     * Confirm that partition states are restored. This method should be called after restoring state of all partitions
+     * in group using {@link #restoreStateOfPartition(int, Integer)}.
+     */
+    void confirmPartitionStatesRestored();
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManager.java:112]: I am not sure - current method name completely matches with what it does

**Review by ibessonov — APPROVED** [review]

**Pull request closed without merging**